src/events/io.rs

```diff
@@ -1 +1,187 @@
-// TODO(benjaminch): implement
+use crate::cloud_provider::io::Kind;
+use crate::errors::io::EngineError;
+use crate::events;
+use serde_derive::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize)]
+#[serde(tag = "type", content = "event")]
+#[serde(rename_all = "lowercase")]
+pub enum EngineEvent {
+    Error(EngineError),
+    Waiting(EventDetails, EventMessage),
+    Deploying(EventDetails, EventMessage),
+    Pausing(EventDetails, EventMessage),
+    Deleting(EventDetails, EventMessage),
+    Deployed(EventDetails, EventMessage),
+    Paused(EventDetails, EventMessage),
+    Deleted(EventDetails, EventMessage),
+}
+
+impl From<events::EngineEvent> for EngineEvent {
+    fn from(event: events::EngineEvent) -> Self {
+        match event {
+            events::EngineEvent::Error(e) => EngineEvent::Error(EngineError::from(e)),
+            events::EngineEvent::Waiting(d, m) => EngineEvent::Waiting(EventDetails::from(d), EventMessage::from(m)),
+            events::EngineEvent::Deploying(d, m) => {
+                EngineEvent::Deploying(EventDetails::from(d), EventMessage::from(m))
+            }
+            events::EngineEvent::Pausing(d, m) => EngineEvent::Pausing(EventDetails::from(d), EventMessage::from(m)),
+            events::EngineEvent::Deleting(d, m) => EngineEvent::Deleting(EventDetails::from(d), EventMessage::from(m)),
+            events::EngineEvent::Deployed(d, m) => EngineEvent::Deployed(EventDetails::from(d), EventMessage::from(m)),
+            events::EngineEvent::Paused(d, m) => EngineEvent::Paused(EventDetails::from(d), EventMessage::from(m)),
+            events::EngineEvent::Deleted(d, m) => EngineEvent::Deleted(EventDetails::from(d), EventMessage::from(m)),
+        }
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+#[serde(rename_all = "lowercase")]
+pub struct EventMessage {
+    raw: String,
+    safe: Option<String>,
+}
+
+impl From<events::EventMessage> for EventMessage {
+    fn from(message: events::EventMessage) -> Self {
+        EventMessage {
+            raw: message.raw,
+            safe: message.safe,
+        }
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+#[serde(rename_all = "lowercase")]
+pub enum Stage {
+    Infrastructure(InfrastructureStep),
+    Environment(EnvironmentStep),
+}
+
+impl From<events::Stage> for Stage {
+    fn from(stage: events::Stage) -> Self {
+        match stage {
+            events::Stage::Infrastructure(step) => Stage::Infrastructure(InfrastructureStep::from(step)),
+            events::Stage::Environment(step) => Stage::Environment(EnvironmentStep::from(step)),
+        }
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+#[serde(rename_all = "lowercase")]
+pub enum InfrastructureStep {
+    Instantiate,
+    Create,
```

Comment:
What is the difference between Instantiate and create? Isn't it deploy missing?

Reply:
Instanciate is when we purely pop new K8s object (constructor). This  can be called for any step.

In my understanding there is no such thing as deploy, everything is a deploy for infra, it's either create, upgrade or delete.

I might miss some statuses such as Resume or stuff like that I will add those along when calling this new logger.